### PR TITLE
Implement find references and rename, and improve goto definition

### PR DESCRIPTION
**Pull request by nguerrera**
* Add support for find all references and rename to the language server and playground.
* Make goto definition work on namespaces, usings, and decorators.

Fix https://github.com/Azure/cadl-azure/issues/612
Fix https://github.com/Azure/cadl-azure/issues/613

### CODE REVIEW

File: packages/compiler/core/parser.ts [hunk 1]
```diff
@@ -2421,3 +2423,42 @@ function isBlocklessNamespace(node: Node) {
...
+      break;
+    case SyntaxKind.DirectiveExpression:
+      kind = IdentifierKind.Directive;
+      break;
+    default:
+      compilerAssert(
+        (id.parent as any).id === id,
+        "This identifier does not appear to be a declaration. There is probably a missing switch case above."
+      );
+      kind = IdentifierKind.Declaration;
```
Comment (timotheeguerin) [inline packages/compiler/core/parser.ts:2459]: Does this line need to be there, isn't the compileAssert going to crash the program?

File: packages/compiler/core/checker.ts [hunk 2]
```diff
@@ -3042,7 +3082,7 @@ export function createChecker(program: Program): Checker {
   }
 }
 
-function isErrorType(type: Type): type is ErrorType {
+export function isErrorType(type: Type): type is ErrorType {
```
Comment (timotheeguerin) [inline packages/compiler/core/checker.ts:3085]: Seems like we have this helper method already in `decorators.ts along side with isVoidType, isNeverType, should we move them all in the checker or into a new helper file?

File: packages/compiler/core/checker.ts [hunk 3]
```diff
@@ -1088,14 +1091,57 @@ export function createChecker(program: Program): Checker {
...
+          }
+        }
+        sym = resolveTypeReference(ref, resolveDecorator);
+        break;
+      default:
+        const _assertNever: never = kind;
+        compilerAssert(false, "Unreachable");
+    }
+
+    return sym?.symbolSource ? sym.symbolSource : sym;
```
Comment (timotheeguerin) [inline packages/compiler/core/checker.ts:1126]:
```suggestion
    return sym?.symbolSource ?? sym;
```

File: packages/playground/src/services.ts [hunk 4]
```diff
@@ -17,65 +18,116 @@ export function attachServices(host: BrowserHost) {
...
+      return results.map(monacoLocation);
     },
   });
 
-  const lsConfig = serverLib.initialize({
-    capabilities: {},
-    processId: 1,
-    workspaceFolders: [],
-    rootUri: "inmemory://",
+  monaco.languages.registerRenameProvider("cadl", {
```
Comment (timotheeguerin) [inline packages/playground/src/services.ts:108]: Does this work automatically with some default shortcuts or is it only available via a context menu?

**Review by timotheeguerin — APPROVED** [review]
Sweet!

File: packages/compiler/core/parser.ts (hunk 1 above)
Comment (nguerrera) [reply packages/compiler/core/parser.ts:2459]: I would prefer to get this crash then to debug strange behavior that would result if we call this a declaration and it isn't.

File: packages/compiler/core/checker.ts (hunk 2 above)
Comment (nguerrera) [reply packages/compiler/core/checker.ts:3085]: Probably. I didn't end up needing this so I'll revert it to keep the PR on task.

File: packages/playground/src/services.ts (hunk 4 above)
Comment (nguerrera) [reply packages/playground/src/services.ts:108]: There are default shortcuts. F2 to rename works great.

File: packages/compiler/core/checker.ts (hunk 2 above)
Comment (nguerrera) [inline packages/compiler/core/checker.ts:3085]:
```suggestion
function isErrorType(type: Type): type is ErrorType {
```

Comment (nguerrera) [reply packages/compiler/core/checker.ts]: That said, I'm almost sure I've missed some things here, I'll see if I can run this over all the identifiers in our tests.

File: packages/compiler/core/parser.ts (hunk 1 above)
Comment (nguerrera) [reply packages/compiler/core/parser.ts:2459]: That said, I'm almost sure I've missed some things here, I'll see if I can run this over all the identifiers in our tests.

Comment (timotheeguerin) [reply packages/compiler/core/parser.ts:2459]: oh... my bad, didn't see the compilerAssert had a condition and wasn't always on for this switch case

Comment (nguerrera) [reply packages/compiler/core/parser.ts:2459]: Testing found lots of missed cases in projections. I'm changing this to set the kind to "Other" when this condition doesn't hold so we can add support over time.

**Comment by nguerrera** [comment]
@timotheeguerin Mind looking at my changes since your review? They're not 100% trivial.

**Review by timotheeguerin — APPROVED** [review]

**Pull request merged**